src/full/Agda/Utils/Monad.hs

```diff
@@ -169,11 +169,12 @@ dropWhileEndM p (x : xs) = ifNotNullM (dropWhileEndM p xs) (return . (x:)) $ {-e
   ifM (p x) (return []) (return [x])
 
 -- | A ``monadic'' version of @'partition' :: (a -> Bool) -> [a] -> ([a],[a])
-partitionM :: (Functor m, Applicative m) => (a -> m Bool) -> [a] -> m ([a],[a])
-partitionM f [] =
-  pure ([], [])
-partitionM f (x:xs) =
-  (\ b (l, r) -> if b then (x:l, r) else (l, x:r)) <$> f x <*> partitionM f xs
+partitionM :: (Functor m, Applicative m) => (a -> m Bool) -> [a] -> m ([a], [a])
+partitionM f xs =
+  foldr
+    (\x -> (<*>) ((\b (l, r) -> if b then (x : l, r) else (l, x : r)) <$> f x))
+    (pure ([], []))
+    xs
```

Comment:
The refactoring in terms of `foldr` is much less readable than the original recursive definition.

Reply:
I'd first factor out a global utility function (better name welcome...)
````
consToOneOfPair :: a -> Bool -> ([a], [a]) ->  ([a], [a])
consToOneOfPair x b (l, r) = if b then (l, x : r) else (x : l, r)
````
and then look at the two options again...

(Note that I switched the alternatives, since the order on `Bool` has `False < True`. This function just uses `Bool` as implementation of `Fin 2`...)

Reply:
```haskell
consToOneOfPair x b = if b then second (x:) else first (x:)

constToOneOfPair x b = bool second first b (x:)
```
A smarter linter could have produced this:
```haskell
partitionM :: (Functor m, Applicative m) => (a -> m Bool) -> [a] -> m ([a], [a])
partitionM f = foldr $ \ x -> liftA2 (bool (first (x:)) (second (x:))) (f x)
```
or maybe (less eta-reduced):
```
partitionM f = foldr $ \ x  myzs -> bool (first (x:)) (second (x:)) <$> f x <*> myzs
```
But still, why would that be more readable than a honest-to-God recursion?  
As always, depends on taste.  But that is an argument to _not_ throw `hlint` onto an established code base...